src/libstore/s3-binary-cache-store.cc

```diff
@@ -271,34 +277,76 @@ struct S3BinaryCacheStoreImpl : public S3BinaryCacheStore
         const std::string & mimeType,
         const std::string & contentEncoding)
     {
-        auto request =
-            Aws::S3::Model::PutObjectRequest()
-            .WithBucket(bucketName)
-            .WithKey(path);
+        auto stream = std::make_shared<istringstream_nocopy>(data);
 
-        request.SetContentType(mimeType);
+        auto maxThreads = std::thread::hardware_concurrency();
 
-        if (contentEncoding != "")
-            request.SetContentEncoding(contentEncoding);
+        auto executor =
+            std::make_shared<Aws::Utils::Threading::PooledThreadExecutor>(maxThreads);
```

Comment:
This thread pool should be shared between invocations, otherwise e.g. `nix copy` will start cores^2 threads (since it will do multiple copies in parallel). E.g.
```
static std::make_shared<Aws::Utils::Threading::PooledThreadExecutor> = std::make_shared<...>(maxThreads);
```
```

Reply:
@edolstra  I pushed a fix in case you missed it :)